[BUG]: Some contract reads result in a crash

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

```console
🔐 > governance.get_proposal_by_id 0x01
panic: proto: koinos.contracts.governance.get_proposal_by_id_result.value: assigning invalid zero-value message

goroutine 1 [running]:
google.golang.org/protobuf/types/dynamicpb.typecheck(...)
	/Users/sgerbino/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/koinos/protobuf-go@v1.27.2-0.20211026185306-2456c83214fe/types/dynamicpb/dynamic.go:520
google.golang.org/protobuf/types/dynamicpb.(*Message).Set(0xc00049e5c0, {0x1007753a8?, 0xc000864d20}, {{}, 0x100569a80?, 0xc00049e600?, 0x3?})
	/Users/sgerbino/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/koinos/protobuf-go@v1.27.2-0.20211026185306-2456c83214fe/types/dynamicpb/dynamic.go:267 +0x2d0
github.com/koinos/koinos-cli/internal/cli.(*ReadContractCommand).Execute(0xc000128038, {0x10076e9b0, 0xc00012c008}, 0xc0002cec40)
	/private/tmp/koinos-cli-20221205-2404-280ca4/internal/cli/contract_commands.go:261 +0x9ba
github.com/koinos/koinos-cli/internal/cli.(*ParseResults).Interpret(0xc0000120f0, 0xc000167cb0?)
	/private/tmp/koinos-cli-20221205-2404-280ca4/internal/cli/interpreter.go:394 +0xf3
github.com/koinos/koinos-cli/internal/cli.ParseAndInterpret(0x1215177fffff04?, 0xf161311191a1c03?, {0xc000167cb0?, 0x9600?})
	/private/tmp/koinos-cli-20221205-2404-280ca4/internal/cli/interpreter.go:456 +0x3d8
github.com/koinos/koinos-cli/cmd/cli/interactive.(*KoinosPrompt).executor(0x1001168ca?, {0xc000167cb0?, 0x100aa5900?})
	/private/tmp/koinos-cli-20221205-2404-280ca4/cmd/cli/interactive/interactive.go:117 +0x31
github.com/koinos/go-prompt.(*Prompt).Run(0xc0000aba70)
	/Users/sgerbino/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/koinos/go-prompt@v0.0.0-20221201222302-dba4c3542a91/prompt.go:84 +0x755
github.com/koinos/koinos-cli/cmd/cli/interactive.(*KoinosPrompt).Run(0xc0000015e0)
	/private/tmp/koinos-cli-20221205-2404-280ca4/cmd/cli/interactive/interactive.go:125 +0xb4
main.main()
	/private/tmp/koinos-cli-20221205-2404-280ca4/cmd/cli/main.go:117 +0x5ea
```

### Expected behavior

None of these calls should result in a panic.

### Steps to reproduce

_No response_

### Environment

```markdown
- OS:
```


### Anything else?

_No response_

--- a/internal/cli/commands.go
+++ b/internal/cli/commands.go
@@ -956,59 +956,85 @@ func (c *RcLimitCommand) Execute(ctx context.Context, ee *ExecutionEnvironment) 
 	// If no limit given, display current
 	if c.limit == nil {
 		if ee.rcLimit.absolute {
-			result.AddMessage(fmt.Sprintf("Current rc limit: %f", ee.rcLimit.value))
+			decAmount, err := util.SatoshiToDecimal(ee.rcLimit.value, cliutil.KoinPrecision)
+			if err != nil {
+				return nil, err
+			}
+			result.AddMessage(fmt.Sprintf("Current rc limit: %v", decAmount))
 			return result, nil
 		}
 
 		// Otherwise its relative
 		if !ee.IsOnline() || !ee.IsWalletOpen() {
-			result.AddMessage(fmt.Sprintf("Current rc limit: %f%%", ee.rcLimit.value*100))
+			decAmount, err := util.SatoshiToDecimal(ee.rcLimit.value, cliutil.KoinPrecision)
+			resultVal := decimal.NewFromFloat(100).Mul(*decAmount)
+			if err != nil {
+				return nil, err
+			}
+			result.AddMessage(fmt.Sprintf("Current rc limit: %v%%", resultVal))
 			return result, nil
 		}
 
-		limit, err := ee.RPCClient.GetAccountRc(ctx, ee.Key.AddressBytes())
+		amount, err := ee.GetRcLimit(ctx)
 		if err != nil {
 			return nil, err
 		}
 
-		dAmount, err := util.SatoshiToDecimal(limit, cliutil.KoinPrecision)
+		decAmount, err := util.SatoshiToDecimal(amount, cliutil.KoinPrecision)
 		if err != nil {
 			return nil, err
 		}
 
-		f, _ := dAmount.Mul(decimal.NewFromFloat(ee.rcLimit.value)).Float64()
-		result.AddMessage(fmt.Sprintf("Current rc limit: %f%% (%f)", ee.rcLimit.value*100, f))
+		decLimit, err := util.SatoshiToDecimal(ee.rcLimit.value, cliutil.KoinPrecision)
+		if err != nil {
+			return nil, err
+		}
+
+		result.AddMessage(fmt.Sprintf("Current rc limit: %v%% (%v)", decLimit.Mul(decimal.NewFromInt(100)), decAmount))
 		return result, nil
 	}
 
 	// Otherwise we are setting the limit
 	s := *c.limit
 	if s[len(s)-1] == '%' {
-		res, err := strconv.ParseFloat(s[:len(s)-1], 64)
+		res, err := decimal.NewFromString(s[:len(s)-1])
 		if err != nil {
 			return nil, err
 		}
 
 		// Check bounds
-		if res < 0 || res > 100 {
+		if res.LessThan(decimal.NewFromInt(0)) || res.GreaterThan(decimal.NewFromInt(100)) {
 			return nil, fmt.Errorf("%w: percentage rc limit must be between 0%% and 100%%", cliutil.ErrInvalidParam)
 		}
 
-		ee.rcLimit.value = res / 100.0
+		// Convert to decimal
+		resFrac := res.Div(decimal.NewFromInt(100))
+		val, err := util.DecimalToSatoshi(&resFrac, cliutil.KoinPrecision)
+		if err != nil {
+			return nil, err
+		}
+
+		ee.rcLimit.value = val
 		ee.rcLimit.absolute = false
-		result.AddMessage(fmt.Sprintf("Set rc limit to %f%%", res))
+		result.AddMessage(fmt.Sprintf("Set rc limit to %v%%", res))
 		return result, nil
 	}
 
 	// Otherwise we are setting the absolute limit
-	res, err := strconv.ParseFloat(s, 64)
+	res, err := decimal.NewFromString(s)
 	if err != nil {
 		return nil, err
 	}
 
-	ee.rcLimit.value = res
+	// Convert to satoshi
+	val, err := util.DecimalToSatoshi(&res, cliutil.KoinPrecision)
+	if err != nil {
+		return nil, err
+	}
+
+	ee.rcLimit.value = val
 	ee.rcLimit.absolute = true
-	result.AddMessage(fmt.Sprintf("Set rc limit to %f", res))
+	result.AddMessage(fmt.Sprintf("Set rc limit to %v", res))
 
 	return result, nil
 }

--- a/internal/cli/contract_commands.go
+++ b/internal/cli/contract_commands.go
@@ -204,8 +204,22 @@ func (c *ReadContractCommand) Execute(ctx context.Context, ee *ExecutionEnvironm
 
 	er := NewExecutionResult()
 
+	DecodeMessageBytes(dMsg, md)
+
+	b, err := text.MarshalPretty(dMsg)
+	if err != nil {
+		return nil, err
+	}
+
+	er.AddMessage(string(b))
+
+	return er, nil
+}
+
+func DecodeMessageBytes(dMsg *dynamicpb.Message, md protoreflect.MessageDescriptor) (protoreflect.Message, error) {
 	l := md.Fields().Len()
 	for i := 0; i < l; i++ {
+		modified := false
 		fd := md.Fields().Get(i)
 		value := dMsg.Get(fd)
 
@@ -222,35 +236,30 @@ func (c *ReadContractCommand) Execute(ctx context.Context, ee *ExecutionEnvironm
 
 				switch koinos.BytesType(enum) {
 				case koinos.BytesType_HEX, koinos.BytesType_BLOCK_ID, koinos.BytesType_TRANSACTION_ID:
-					b = []byte(hex.EncodeToString(value.Bytes()))
-					if len(b) == 0 && len(value.Bytes()) != 0 {
-						err = fmt.Errorf("error encoding hex")
+					b, err = hex.DecodeString(string(value.Bytes()))
+					if err != nil {
+						return nil, err
 					}
 				case koinos.BytesType_BASE58, koinos.BytesType_CONTRACT_ID, koinos.BytesType_ADDRESS:
-					b = []byte(base58.Encode(value.Bytes()))
-					if len(b) == 0 && len(value.Bytes()) != 0 {
-						err = fmt.Errorf("error encoding base58")
-					}
+					b = base58.Decode(string(value.Bytes()))
+
 				case koinos.BytesType_BASE64:
 					fallthrough
 				default:
-					b = []byte(base64.URLEncoding.EncodeToString(value.Bytes()))
-					if len(b) == 0 && len(value.Bytes()) != 0 {
-						err = fmt.Errorf("error encoding base64")
+					b, err = base64.URLEncoding.DecodeString(string(value.Bytes()))
+					if err != nil {
+						return nil, err
 					}
 				}
 			} else {
-				b = []byte(base64.URLEncoding.EncodeToString(value.Bytes()))
-				if len(b) == 0 && len(value.Bytes()) != 0 {
-					err = fmt.Errorf("error encoding base64")
+				b, err = base64.URLEncoding.DecodeString(string(value.Bytes()))
+				if err != nil {
+					return nil, err
 				}
 			}
 
-			if err != nil {
-				return nil, err
-			}
-
 			value = protoreflect.ValueOfBytes(b)
+			modified = true
 		}
 
 		if fd.IsList() && value.List().Len() == 0 {
@@ -258,17 +267,12 @@ func (c *ReadContractCommand) Execute(ctx context.Context, ee *ExecutionEnvironm
 		}
 
 		// Set the value on the message
-		dMsg.Set(fd, value)
+		if modified {
+			dMsg.Set(fd, value)
+		}
 	}
 
-	b, err := text.MarshalPretty(dMsg)
-	if err != nil {
-		return nil, err
-	}
-
-	er.AddMessage(string(b))
-
-	return er, nil
+	return dMsg, nil
 }
 
 // ----------------------------------------------------------------------------

--- a/internal/cli/contract_commands.go
+++ b/internal/cli/contract_commands.go
@@ -204,7 +204,10 @@ func (c *ReadContractCommand) Execute(ctx context.Context, ee *ExecutionEnvironm
 
 	er := NewExecutionResult()
 
-	DecodeMessageBytes(dMsg, md)
+	err = DecodeMessageBytes(dMsg, md)
+	if err != nil {
+		return nil, err
+	}
 
 	b, err := text.MarshalPretty(dMsg)
 	if err != nil {
@@ -216,7 +219,7 @@ func (c *ReadContractCommand) Execute(ctx context.Context, ee *ExecutionEnvironm
 	return er, nil
 }
 
-func DecodeMessageBytes(dMsg *dynamicpb.Message, md protoreflect.MessageDescriptor) (protoreflect.Message, error) {
+func DecodeMessageBytes(dMsg *dynamicpb.Message, md protoreflect.MessageDescriptor) error {
 	l := md.Fields().Len()
 	for i := 0; i < l; i++ {
 		modified := false
@@ -238,7 +241,7 @@ func DecodeMessageBytes(dMsg *dynamicpb.Message, md protoreflect.MessageDescript
 				case koinos.BytesType_HEX, koinos.BytesType_BLOCK_ID, koinos.BytesType_TRANSACTION_ID:
 					b, err = hex.DecodeString(string(value.Bytes()))
 					if err != nil {
-						return nil, err
+						return err
 					}
 				case koinos.BytesType_BASE58, koinos.BytesType_CONTRACT_ID, koinos.BytesType_ADDRESS:
 					b = base58.Decode(string(value.Bytes()))
@@ -248,13 +251,13 @@ func DecodeMessageBytes(dMsg *dynamicpb.Message, md protoreflect.MessageDescript
 				default:
 					b, err = base64.URLEncoding.DecodeString(string(value.Bytes()))
 					if err != nil {
-						return nil, err
+						return err
 					}
 				}
 			} else {
 				b, err = base64.URLEncoding.DecodeString(string(value.Bytes()))
 				if err != nil {
-					return nil, err
+					return err
 				}
 			}
 
@@ -272,7 +275,7 @@ func DecodeMessageBytes(dMsg *dynamicpb.Message, md protoreflect.MessageDescript
 		}
 	}
 
-	return dMsg, nil
+	return nil
 }
 
 // ----------------------------------------------------------------------------

--- a/internal/cli/interpreter.go
+++ b/internal/cli/interpreter.go
@@ -12,7 +12,6 @@ import (
 	"github.com/koinos/koinos-cli/internal/cliutil"
 	"github.com/koinos/koinos-proto-golang/koinos/protocol"
 	util "github.com/koinos/koinos-util-golang"
-	"github.com/shopspring/decimal"
 )
 
 // Command execution code
@@ -55,7 +54,7 @@ func (er *ExecutionResult) Print() {
 }
 
 type rcInfo struct {
-	value    float64
+	value    uint64
 	absolute bool
 }
 
@@ -86,7 +85,7 @@ func NewExecutionEnvironment(rpcClient *cliutil.KoinosRPCClient, parser *Command
 		Contracts: make(map[string]*ContractInfo),
 		Session:   &TransactionSession{},
 		nonceMap:  make(map[string]*nonceInfo),
-		rcLimit:   rcInfo{value: 1.0, absolute: false},
+		rcLimit:   rcInfo{value: 100000000, absolute: false},
 		payer:     SelfPayer,
 		chainID:   AutoChainID,
 		nonceMode: AutoNonce,
@@ -184,14 +183,7 @@ func (ee *ExecutionEnvironment) GetChainID(ctx context.Context) ([]byte, error) 
 // GetRcLimit returns the current RC limit
 func (ee *ExecutionEnvironment) GetRcLimit(ctx context.Context) (uint64, error) {
 	if ee.rcLimit.absolute {
-		dAmount := decimal.NewFromFloat(ee.rcLimit.value)
-
-		val, err := util.DecimalToSatoshi(&dAmount, cliutil.KoinPrecision)
-		if err != nil {
-			return 0, fmt.Errorf("%w: %s", cliutil.ErrInvalidAmount, err.Error())
-		}
-
-		return val, nil
+		return ee.rcLimit.value, nil
 	}
 
 	// else it's relative
@@ -200,8 +192,23 @@ func (ee *ExecutionEnvironment) GetRcLimit(ctx context.Context) (uint64, error) 
 		return 0, err
 	}
 
-	val := uint64(float64(limit) * ee.rcLimit.value)
-	return val, nil
+	decLimit, err := util.SatoshiToDecimal(limit, 8)
+	if err != nil {
+		return 0, err
+	}
+
+	decVal, err := util.SatoshiToDecimal(ee.rcLimit.value, 8)
+	if err != nil {
+		return 0, err
+	}
+
+	decResult := decLimit.Mul(*decVal)
+	res, err := util.DecimalToSatoshi(&decResult, 8)
+	if err != nil {
+		return 0, err
+	}
+
+	return res, nil
 }
 
 // SubmitTransaction is a utility function to submit a transaction from a command

